File: research-graph/scripts/medhorizon-gateway.ts
#!/usr/bin/env bun
/**
 * MedHorizon + Research Graph companion gateway.
 *
 * Proxies MedHorizon's web UI and injects the Research Graph sidebar-card
 * script into HTML responses. Does NOT modify MedHorizon source.
 *
 * Usage:
 *   bun research-graph/scripts/medhorizon-gateway.ts
 *
 * Env:
 *   MEDHORIZON_ORIGIN   default http://127.0.0.1:4444
 *   RESEARCH_GRAPH_API  default http://127.0.0.1:8000
 *   GATEWAY_PORT        default 5199
 */

const origin = (process.env.MEDHORIZON_ORIGIN || "http://127.0.0.1:4444").replace(/\/$/, "")
const api = (process.env.RESEARCH_GRAPH_API || "http://127.0.0.1:8000").replace(/\/$/, "")
const port = Number(process.env.GATEWAY_PORT || 5199)
const snippet = `<script src="${api}/embed/sidebar-card.js" data-rg-api="${api}" defer></script>`

function inject(html: string) {
  if (html.includes("sidebar-card.js") || html.includes("data-rg-card")) return html
  if (html.includes("</head>")) return html.replace("</head>", `${snippet}\n</head>`)
  if (html.includes("</body>")) return html.replace("</body>", `${snippet}\n</body>`)
  return html + snippet
}

const server = Bun.serve({
  port,
  hostname: "127.0.0.1",
  async fetch(req) {
    const url = new URL(req.url)
    const target = `${origin}${url.pathname}${url.search}`
    const headers = new Headers(req.headers)
    headers.delete("host")
    headers.delete("accept-encoding")

    const upstream = await fetch(target, {
      method: req.method,
      headers,
      body: req.method === "GET" || req.method === "HEAD" ? undefined : await req.arrayBuffer(),
      redirect: "manual",
    })

    const type = upstream.headers.get("content-type") || ""
    const out = new Headers(upstream.headers)
    out.delete("content-encoding")
    out.delete("content-length")

    if (type.includes("text/html")) {
      const html = inject(await upstream.text())
      return new Response(html, { status: upstream.status, headers: out })
    }

    return new Response(upstream.body, { status: upstream.status, headers: out })
  },
})

console.log(`Research Graph gateway → MedHorizon ${origin}`)
console.log(`Open http://127.0.0.1:${server.port}  (injects sidebar card from ${api})`)

File: research-graph/scripts/medhorizon-gateway.py
#!/usr/bin/env python3
"""MedHorizon companion gateway — inject Research Graph sidebar card (no core edits)."""

from __future__ import annotations

import os
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.request import Request, urlopen

ORIGIN = os.environ.get("MEDHORIZON_ORIGIN", "http://127.0.0.1:4444").rstrip("/")
API = os.environ.get("RESEARCH_GRAPH_API", "http://127.0.0.1:8000").rstrip("/")
PORT = int(os.environ.get("GATEWAY_PORT", "5199"))
SNIPPET = f'<script src="{API}/embed/sidebar-card.js" data-rg-api="{API}" defer></script>'


def inject(html: str) -> str:
    if "sidebar-card.js" in html or "data-rg-card" in html:
        return html
    if "</head>" in html:
        return html.replace("</head>", f"{SNIPPET}\n</head>", 1)
    if "</body>" in html:
        return html.replace("</body>", f"{SNIPPET}\n</body>", 1)
    return html + SNIPPET


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):  # noqa: N802
        self._proxy()

    def do_POST(self):  # noqa: N802
        self._proxy()

    def log_message(self, fmt: str, *args) -> None:
        print(f"[gateway] {self.address_string()} {fmt % args}")

    def _proxy(self) -> None:
        target = f"{ORIGIN}{self.path}"
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length) if length else None
        headers = {k: v for k, v in self.headers.items() if k.lower() not in {"host", "accept-encoding"}}
        req = Request(target, data=body, headers=headers, method=self.command)
        with urlopen(req, timeout=30) as res:  # noqa: S310 — local loopback proxy
            data = res.read()
            ctype = res.headers.get("Content-Type", "")
            out = data
            if "text/html" in ctype:
                out = inject(data.decode("utf-8", errors="ignore")).encode("utf-8")
            self.send_response(res.status)
            for key, value in res.headers.items():
                if key.lower() in {"content-encoding", "content-length", "transfer-encoding"}:
                    continue
                self.send_header(key, value)
            self.send_header("Content-Length", str(len(out)))
            self.end_headers()
            self.wfile.write(out)


if __name__ == "__main__":
    print(f"Research Graph gateway → MedHorizon {ORIGIN}")
    print(f"Open http://127.0.0.1:{PORT}  (injects sidebar card from {API})")
    ThreadingHTTPServer(("127.0.0.1", PORT), Handler).serve_forever()
